### PR TITLE
Resolve #871 -- Upgrade to .NET Core

### LINKED ISSUE #869: Migrate projects to new csproj format

### LINKED ISSUE #871: Use .NET Core
This will allow the server to be ran on Linux and MacOS

### PR DESCRIPTION
**Pull request by danil179**
This PR continues #870 PR
Fixing travis build
Then upgrading to .NET Core

Resolve #869
Resolve #871

IMPORTANT:
To reserve credit to pipe01 after squash/merge I think you have to add Co-authored-by

### CODE REVIEW

File: GameServerConsole/GameServerConsole.csproj [hunk 1]
```diff
@@ -1,126 +1,102 @@
...
-  <PropertyGroup Condition="'$(OS)' != 'Unix'">
-    <_CopyCommand>xcopy /s /y /q /d</_CopyCommand>
-  </PropertyGroup>
-  <PropertyGroup Condition="'$(OS)' == 'Unix'">
-    <_CopyCommand Condition="'$(_IsMac)' == 'false'">cp -r -u</_CopyCommand>
-    <_CopyCommand Condition="'$(_IsMac)' == 'true'">cp -Rf</_CopyCommand>
-  </PropertyGroup>
-  <Target Name="AfterBuild">
-    <Exec Command="$(_CopyCommand) &quot;$(_LibSourceDirectory)&quot; &quot;$(_LibTargetDirectory)&quot;" />
-    <Exec Command="$(_CopyCommand) &quot;$(_ENetSharpLeagueSourceDirectory)&quot; &quot;$(_ENetSharpLeagueTargetDirectory)&quot;" />
```
Comment (danil179) [inline GameServerConsole/GameServerConsole.csproj:144]: This part should never have been deleted. This part copies the `libenet.dll` and other libs as well settings .

File: GameServerConsole/Settings/GameInfo.json.template [hunk 2]
```diff
@@ -54,7 +54,7 @@
         "COOLDOWNS_ENABLED": false,
         "CHEATS_ENABLED": true,
         "MINION_SPAWNS_ENABLED": false,
-        "CONTENT_PATH": "../../../Content",
+        "CONTENT_PATH": "../../../../Content",
```
Comment (danil179) [inline GameServerConsole/Settings/GameInfo.json.template]: Check if that part causes crash.

File: GameServerConsoleTests/packages.config [hunk 3]
```diff
@@ -1,17 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<packages>
-  <package id="CommandLineParser" version="2.3.0" targetFramework="net462" />
```
Comment (danil179) [inline GameServerConsoleTests/packages.config]: Check if it is really not needed.

File: GameServerCore/GameServerCore.csproj [hunk 4]
```diff
@@ -1,264 +1,24 @@
...
     <DebugType>full</DebugType>
-    <Optimize>false</Optimize>
-    <OutputPath>bin\Debug\</OutputPath>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
-    <ErrorReport>prompt</ErrorReport>
-    <WarningLevel>4</WarningLevel>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
     <DebugType>pdbonly</DebugType>
-    <Optimize>true</Optimize>
```
Comment (danil179) [inline GameServerCore/GameServerCore.csproj:27]: Do we really don't want to optimize or maybe is it the default?

File: GameServerLibTests/GameServerLibTests.csproj [hunk 5]
```diff
@@ -1,107 +1,37 @@
...
-    <VisualStudioVersion Condition="'$(VisualStudioVersion)' == ''">10.0</VisualStudioVersion>
     <VSToolsPath Condition="'$(VSToolsPath)' == ''">$(MSBuildExtensionsPath32)\Microsoft\VisualStudio\v$(VisualStudioVersion)</VSToolsPath>
     <ReferencePath>$(ProgramFiles)\Common Files\microsoft shared\VSTT\$(VisualStudioVersion)\UITestExtensionPackages</ReferencePath>
     <IsCodedUITest>False</IsCodedUITest>
     <TestProjectType>UnitTest</TestProjectType>
-    <TargetFrameworkProfile />
+    <AssemblyTitle>GameServerTests</AssemblyTitle>
+    <Company>Hewlett-Packard Company</Company>
+    <Product>GameServerTests</Product>
+    <Copyright>Copyright © Hewlett-Packard Company 2016</Copyright>
```
Comment (danil179) [inline GameServerLibTests/GameServerLibTests.csproj]: I don't think Hewlett-Packard Company is somehow related. Change this.

File: appveyor.yml [hunk 6]
```diff
@@ -11,8 +11,10 @@ build:
   verbosity: minimal
 
 test_script:
-  - .\packages\OpenCover.4.6.519\tools\OpenCover.Console.exe -register:user -target:"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\MSTest.exe" -targetargs:"/noresults /noisolation /testcontainer:"".\GameServerAppTests\bin\Debug\GameServerAppTests.dll" -filter:"+[GameServerApp]*  -[GameServerApp]GameServerApp.Properties.*" -excludebyattribute:*.ExcludeFromCodeCoverage* -hideskipped:All -output:.\GameServerApp_coverage.xml
-  - .\packages\OpenCover.4.6.519\tools\OpenCover.Console.exe -register:user -target:"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\MSTest.exe" -targetargs:"/noresults /noisolation /testcontainer:"".\GameServerLibTests\bin\Debug\GameServerLibTests.dll" -filter:"+[GameServerLib]*  -[GameServerLib]GameServerLib.Properties.*" -excludebyattribute:*.ExcludeFromCodeCoverage* -hideskipped:All -output:.\GameServerLib_coverage.xml
+  - curl -fsSL -o opencover.zip https://github.com/OpenCover/opencover/releases/download/4.7.922/opencover.4.7.922.zip
```
Comment (danil179) [inline appveyor.yml:14]: Seems better approach as updating this will be easier now. Still, I want to find version independent approach.

**Review by danil179 — COMMENTED** [review]
Changes requested.

Apply to #870 as well.

File: GameServerConsole/GameServerConsole.csproj (hunk 1 above)
Comment (danil179) [reply GameServerConsole/GameServerConsole.csproj:144]: I see that it was implemented, but seems that Travis fail to understand this on Unix.

**Comment by danil179** [comment]
Again, tests with relative directories (`..\..\..\..\Content` for item tests) fail. I think this is because `OpenCover` is taken as the current directory.

File: appveyor.yml (hunk 6 above)
Comment (MythicManiac) [reply appveyor.yml:14]: Version locking is fine in this case

**Comment by danil179** [comment]
Project now runs without any special problems. I'll now try to change to .NET core.

**Comment by danil179** [comment]
.NET Core compiles locally. There are major problems with DLL importing.

**Comment by danil179** [comment]
Runtime error:
```
System.BadImageFormatException: 'An attempt was made to load a program with an incorrect format. (0x8007000B)'
```

**Comment by danil179** [comment]
Everything works except log4net logger. This logger seems to ignore every settings for printing, but yet recognize it. When I run it on debug mode I see some logging and then log seems to stop.

**Comment by danil179** [comment]
@MythicManiac It should be safe to squash&merge now, be aware that we need to add `Co-authored-by` that includes pipe01 as also an author of this commit.

EDIT: some tests are now broken cause path changed again

**Comment by danil179** [comment]
The tests of ServerConsole get skipped as something in x86 causes problems.

**Comment by danil179** [comment]
@MythicManiac I think it is safe to merge now

File: GameServerConsole/GameServerConsole.csproj [hunk 7]
```diff
@@ -1,146 +1,68 @@
...
-      <Private>True</Private>
-      <Private>True</Private>
-    </Reference>
-    <Reference Include="System.Xml.Linq" />
-    <Reference Include="System.Data.DataSetExtensions" />
-    <Reference Include="System.Data" />
-    <Reference Include="System.Net.Http" />
-    <Reference Include="System.Xml" />
+    <PackageReference Include="CommandLineParser" Version="2.6.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.2" />
```
Comment (HarmfulBreeze) [inline GameServerConsole/GameServerConsole.csproj:18]: isn't log4net missing here? (prob not, just mentioning it to be sure)

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs [hunk 8]
```diff
@@ -25,7 +24,7 @@ public class BaseTurret : ObjAiBase, IBaseTurret
             uint netId = 0
         ) : base(game, model, new Stats.Stats(), 50, x, y, 1200, netId)
         {
-            ParentNetId = Crc32Algorithm.Compute(Encoding.UTF8.GetBytes(name)) | 0xFF000000;
+            ParentNetId = Force.Crc32.Crc32Algorithm.Compute(Encoding.UTF8.GetBytes(name)) | 0xFF000000;
```
Comment (HarmfulBreeze) [inline GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs]: not sure why this was inlined

**Review by HarmfulBreeze — COMMENTED** [review]
Didn't see anything out of place really, if it works, it probably should be merged.

File: GameServerConsole/GameServerConsole.csproj (hunk 7 above)
Comment (danil179) [reply GameServerConsole/GameServerConsole.csproj:18]: log4net isn't used in GameServerConsole (it is in GameServerLib).

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs (hunk 8 above)
Comment (danil179) [reply GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs]: VS marked the `using` line as problematic for no reason so I just copied it here. I don't mind changing it back if VS do no problems with that.

File: GameServerLibTests/Content/TestContentCollection.cs [hunk 9]
```diff
@@ -1,33 +0,0 @@
...
-namespace GameServerTests
-{
-    [TestClass]
-    public class TestContentCollection
-    {
-        [TestMethod]
-        [DeploymentItem("Content", "Content")]
-        public void TestItemContentCollection()
-        {
-            var collection = ItemContentCollection.LoadItemsFrom("Content/Data/LeagueSandbox-Default/Items");
```
Comment (danil179) [inline GameServerLibTests/Content/TestContentCollection.cs:16]: This file probably got deleted by mistake. I'll upload it back.

File: .travis.yml [hunk 10]
```diff
@@ -1,13 +1,12 @@
...
+mono: none
+dotnet: 3.0
 
 install:
-    - wget https://dist.nuget.org/win-x86-commandline/latest/nuget.exe
-    - sudo mono nuget.exe restore GameServer.sln
+    - dotnet restore
 script:
-    - xbuild /p:Configuration=Debug GameServer.sln
+    - dotnet build
```
Comment (danil179) [inline .travis.yml:10]: Add testing here seems good to me.

**Comment by danil179** [comment]
Travis can't run tests without stalling: https://github.com/microsoft/vstest/issues/2080

File: .travis.yml (hunk 10 above)
Comment (danil179) [reply .travis.yml:10]: https://github.com/LeagueSandbox/GameServer/pull/906#issuecomment-546640898

File: GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs (hunk 8 above)
Comment (HarmfulBreeze) [reply GameServerLib/GameObjects/AttackableUnits/AI/BaseTurret.cs]: Yeah we'd rather have one more using directive than a line that's harder to read

**Review by HarmfulBreeze — APPROVED** [review]
👍

**Comment by moonshadow565** [comment]
What about instead of ``<TargetFramework>netcoreapp3.0</TargetFramework>`` target ``<TargetFramework>netstandard2.1</TargetFramework>``  instead ?

**Pull request merged**